### PR TITLE
feat: snapshot restore 

### PR DESCRIPTION
**Pull request by lmyslinski**

### CODE REVIEW

File: pulumi/volumeReaper/src/volumeReaper.ts [hunk 1]
```diff
@@ -12,6 +12,7 @@ interface StatefulSet extends k8s.V1StatefulSet {
 interface VolumeSnapshot extends Required<k8s.KubernetesObject> {
   metadata: {
     name: string
+    // the actual type is string but the parent object has this messed up
```
Comment (kaladinlight) [inline pulumi/volumeReaper/src/volumeReaper.ts:15]: yeah looking more into this the k8s client has a object serializer for all of the different resources, but with volume snapshots being a crd, there is no defined resource resulting in no deserialization of string -> Date.

we should probably add our own with a comment:
```ts
// no ObjectSerializer exists for VolumeSnapshot CRD, manually deserialize necessary values
const items = body.items.map(item => {
  item.metadata.creationTimestamp = new Date(item.metadata.creationTimestamp)
  return item
})
```

Comment (kaladinlight) [reply pulumi/volumeReaper/src/volumeReaper.ts:15]: Then we can assume whenever we are interacting with `creationTimestamp` thereafter, it is actually a `Date`

File: pulumi/src/volumeSnapshotClient.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import * as k8sClient from '@kubernetes/client-node'
+
+export interface VolumeSnapshot extends Required<k8sClient.KubernetesObject> {
```
Comment (kaladinlight) [inline pulumi/src/volumeSnapshotClient.ts:3]: Would be nice to leverage existing types over copy pasta, but the project structure has become a little jank. I am going to lift all of the node config to top level so we can structure this a bit better and leverage eslint/prettier throughout as intended

File: node/package.json [hunk 3]
```diff
@@ -25,6 +25,7 @@
     "prepare": "cd .. && husky install"
   },
   "dependencies": {
+    "@kubernetes/client-node": "^0.18.1",
```
Comment (kaladinlight) [inline node/package.json:28]: with restructure, we should only need this dependency declared long term

File: node/coinstacks/litecoin/pulumi/index.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import { readFileSync } from 'fs'
 import * as k8s from '@pulumi/kubernetes'
 import { deployApi, createService, deployStatefulService, getConfig, Service } from '../../../../pulumi'
 import { api } from '../../../pulumi'
+import { VolumeSnapshotClient } from '../../../../pulumi/src/volumeSnapshotClient'
```
Comment (kaladinlight) [inline node/coinstacks/litecoin/pulumi/index.ts:6]: export from `index.ts` so we can import along with the rest on line 4

File: pulumi/src/pvcResolver.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+interface PVCFromBackup extends k8s.types.input.core.v1.PersistentVolumeClaim {
+  datasource: {
+    name: string;
+    kind: string;
+    apiGroup: string;
+  }
+}
+
+export const getVolumeClaimTemplates = (serviceName: string, storageSize: string, snapshots?: VolumeSnapshot[]): Array<k8s.types.input.core.v1.PersistentVolumeClaim> => {
```
Comment (kaladinlight) [inline pulumi/src/pvcResolver.ts:12]: alternate solution:
  - filter includes both checks
  - snapshots can be ordered by snapshot client to remove extra sort
  - use `PersistentVolumeClaim` class
  - spread in `dataSource` if there is a valid snapshot found
  - use snapshot contents for `dataSource`
```ts
export const getVolumeClaimTemplates = (
  serviceName: string,
  storageSize: string,
  snapshots: VolumeSnapshot[]
): Array<k8s.types.input.core.v1.PersistentVolumeClaim> => {
  const snapshot = snapshots.filter(snapshot => snapshot.metadata.name.includes(serviceName) && snapshot.status.readyToUse)[0]

  const pvc: k8s.types.input.core.v1.PersistentVolumeClaim = {
    metadata: {
      name: `data-${serviceName}`,
    },
    spec: {
      accessModes: ['ReadWriteOnce'],
      storageClassName: 'ebs-csi-gp2',
      resources: {
        requests: {
          storage: storageSize,
        },
      },
      ...(snapshot && {
        dataSource: {
          name: snapshot.metadata.name,
          kind: snapshot.kind,
          apiGroup: snapshot.apiVersion.split('/')[0],
        },
      }),
    },
  }

  return [pvc]
}
```

File: pulumi/src/volumeSnapshotClient.ts (hunk 2 above)
Comment (lmyslinski) [reply pulumi/src/volumeSnapshotClient.ts:3]: I have that as a 4th step or so - ideally to pull in all Pulumi stuff into a single tree, separate from the coinstack implementations

Comment (kaladinlight) [reply pulumi/src/volumeSnapshotClient.ts:3]: yeah full separation as a later step, have a poc of just pulling node workspaces up one level and should make it quite a bit cleaner

Comment (lmyslinski) [reply pulumi/src/volumeSnapshotClient.ts:3]: I'll keep it in here for now, we'll need to marry the reaper module with the rest to share the import

File: pulumi/src/pvcResolver.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
+import * as k8s from '@pulumi/kubernetes'
+import { VolumeSnapshot } from './volumeSnapshotClient';
+
+interface PVCFromBackup extends k8s.types.input.core.v1.PersistentVolumeClaim {
+  datasource: {
```
[the remark this answers is not in the document]
Comment (lmyslinski) [reply pulumi/src/pvcResolver.ts:5]: With the refactor from below we can skip the type

File: pulumi/src/pvcResolver.ts (hunk 5 above)
Comment (lmyslinski) [reply pulumi/src/pvcResolver.ts:12]: Nice, much cleaner

File: pulumi/volumeReaper/src/volumeReaper.ts (hunk 1 above)
Comment (lmyslinski) [reply pulumi/volumeReaper/src/volumeReaper.ts:15]: Added

File: node/package.json (hunk 3 above)
Comment (lmyslinski) [reply node/package.json:28]: Once we restructure yeah, sadly now it needs to stay

File: pulumi/volumeReaper/tsconfig.json [hunk 7]
```diff
@@ -2,7 +2,7 @@
   "extends": "../../node/tsconfig.json",
   "compilerOptions": {
     "outDir": "dist",
-    "types": ["node"]
+    "types": ["node"],
```
Comment (kaladinlight) [inline pulumi/volumeReaper/tsconfig.json:5]:
```suggestion
    "types": ["node"]
```

File: pulumi/volumeReaper/src/volumeReaper.ts [hunk 8]
```diff
@@ -127,12 +134,10 @@ export class VolumeReaper {
...
       `statefulset=${this.name}`
     )
-
-    const items = snapshots.body.items
+    const items = snapshots.body.items.map(this.deserialize);
     console.log(`Found ${items.length} snapshots for ${this.namespace}.${this.name}`)
-
     // sorted newest -> oldest
-    return items.sort((a, b) => b.metadata.creationTimestamp.getTime() - a.metadata.creationTimestamp.getTime())
+    return items.sort((a, b) => new Date(b.metadata.creationTimestamp).getTime() - new Date(a.metadata.creationTimestamp).getTime())
```
Comment (kaladinlight) [inline pulumi/volumeReaper/src/volumeReaper.ts:140]: now we are deserializing
```suggestion
    return items.sort((a, b) => b.metadata.creationTimestamp.getTime() - a.metadata.creationTimestamp.getTime())
```

File: pulumi/src/volumeSnapshotClient.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+      'snapshot.storage.k8s.io/v1',
+      'VolumeSnapshot',
+      this.namespace,
+      undefined,
+      undefined,
+      undefined,
+      undefined,
+      `statefulset=${assetName}-sts`
+    )
+    return response.body.items.map(this.deserialize);
```
Comment (kaladinlight) [inline pulumi/src/volumeSnapshotClient.ts:51]: sort here for my `getVolumeClaimTemplate` suggestion to work as intended

File: pulumi/src/volumeSnapshotClient.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+      persistentVolumeClaimName: string
+    }
+  },
+  status: {
+    readyToUse: boolean,
+  }
+}
+
+export class VolumeSnapshotClient {
+
```
Comment (kaladinlight) [inline pulumi/src/volumeSnapshotClient.ts:23]:
```suggestion
```

File: node/coinstacks/litecoin/pulumi/index.ts [hunk 11]
```diff
@@ -17,6 +24,7 @@ export = async (): Promise<Outputs> => {
   const asset = config.network !== 'mainnet' ? `${coinstack}-${config.network}` : coinstack
   const outputs: Outputs = {}
   const provider = new k8s.Provider('kube-provider', { kubeconfig })
+  const snapshots = await new VolumeSnapshotClient(kubeconfig, namespace).getVolumeSnapshots(asset)
```
Comment (kaladinlight) [inline node/coinstacks/litecoin/pulumi/index.ts:27]: Add to all coinstacks? or still want to do some extra validation first?

Comment (kaladinlight) [reply node/coinstacks/litecoin/pulumi/index.ts:27]: plan is validate against litecoin and then update remaining coinstacks.

litecoin:
- manual snapshot
- deploy and wait for reaper cron to snapshot
- scale up dev to test restore
- scale back down and update remaining coinstacks

**Review by kaladinlight — APPROVED** [review]

**Pull request merged**